File: crates/libs/lib-core/src/services/planet.rs
use mongodb::bson::{oid::ObjectId, Document};

use crate::{
    errors::CoreError,
    repositories::planet::{CreatePlanetDTO, PlanetRepository, UpdatePlanetDTO},
    schemas::planet::{CreatePlanetSchema, PlanetSchema, UpdatePlanetSchema},
    Result,
};

#[derive(Clone)]
pub struct PlanetService<R: PlanetRepository> {
    repository: R,
}

impl<R: PlanetRepository> PlanetService<R> {
    pub fn new(repository: R) -> Self {
        Self { repository }
    }

    pub async fn create_planet(&self, data: CreatePlanetSchema) -> Result<PlanetSchema> {
        let dto = CreatePlanetDTO {
            system_id: data.system_id,
            threat: data.threat,
        };
        let planet_id = self.repository.create(dto).await?;
        self.find_one_planet(planet_id).await
    }

    pub async fn find_one_planet(&self, oid: ObjectId) -> Result<PlanetSchema> {
        let planet = self.repository.find_one(oid).await?;
        match planet {
            None => Err(CoreError::NotFound),
            Some(planet) => Ok(planet.into()),
        }
    }

    pub async fn find_all_planets(&self, filter: Document) -> Result<Vec<PlanetSchema>> {
        let planets: Vec<PlanetSchema> = self
            .repository
            .find_all(filter)
            .await?
            .iter()
            .map(|v| PlanetSchema::from(v.clone()))
            .collect();

        Ok(planets)
    }

    pub async fn delete_planet(&self, oid: ObjectId) -> Result<()> {
        self.repository.delete(oid).await
    }

    pub async fn update_planet(&self, oid: ObjectId, data: UpdatePlanetSchema) -> Result<()> {
        let dto = UpdatePlanetDTO {
            system_id: data.system_id,
            threat: data.threat,
        };

        self.repository.update(oid, dto).await
    }
}

File: api/internal/service/planet.go
package service

import (
	"astragalaxy/internal/model"
	"astragalaxy/internal/schema"

	"github.com/google/uuid"
)

func (s *Service) CreatePlanet(data schema.CreatePlanetSchema) (*schema.PlanetSchema, error) {
	response, err := s.p.Create(&model.Planet{SystemID: data.SystemID, Name: data.Name, Threat: data.Threat})
	if err != nil {
		return nil, err
	}

	return s.FindOnePlanet(*response)
}

func (s *Service) FindOnePlanet(ID uuid.UUID) (*schema.PlanetSchema, error) {
	response, err := s.p.FindOne(ID)
	if err != nil {
		return nil, err
	}
	schema := schema.PlanetSchema{ID: response.ID, Name: response.Name, SystemID: response.SystemID, Threat: response.Threat}
	return &schema, nil
}

func (s *Service) FindAllPlanets(filter *model.Planet) ([]schema.PlanetSchema, error) {
	response, err := s.p.FindAll(filter)
	if err != nil {
		return nil, err
	}
	var planetSchemas []schema.PlanetSchema

	for _, planet := range response {
		planetSchemas = append(planetSchemas, schema.PlanetSchema{
			ID:       planet.ID,
			Name:     planet.Name,
			SystemID: planet.SystemID,
			Threat:   planet.Threat,
		})
	}

	return planetSchemas, nil
}

func (s *Service) DeletePlanet(ID uuid.UUID) error {
	return s.p.Delete(ID)
}

func (s *Service) UpdatePlanet(ID uuid.UUID, data schema.UpdatePlanetSchema) error {
	planet := model.Planet{
		ID:       ID,
		Name:     data.Name,
		SystemID: data.SystemID,
		Threat:   data.Threat,
	}

	return s.p.Update(&planet)
}
